ui/src/chat/ChatReactions/ReactionsWidget.tsx

```diff
@@ -0,0 +1,123 @@
+import _ from 'lodash';
+import { useMemo, useState } from 'react';
+import * as cn from 'classnames';
+import * as Tabs from '@radix-ui/react-tabs';
+import Avatar from '@/components/Avatar';
+import ShipName from '@/components/ShipName';
+
+function byShortcode(a: [string, string], b: [string, string]) {
+  const aCode = a[1];
+  const bCode = b[1];
+  return aCode.localeCompare(bCode);
+}
+
+function FeelRow({ shortcode, ship }: { shortcode: string; ship: string }) {
+  return (
+    <div className="mb-4 flex w-full items-center justify-between">
+      <div className="flex items-center">
+        <Avatar ship={ship} size="small" />
+        <ShipName className="ml-3" name={ship} showAlias />
+      </div>
+      <em-emoji shortcodes={shortcode} size="1.3em" />
+    </div>
+  );
+}
+
+function FeelTab({
+  tabId,
+  count,
+  selected,
+}: {
+  tabId: string;
+  count: number;
+  selected?: boolean;
+}) {
+  return (
+    <div
+      className={cn(
+        'flex items-center justify-center rounded-lg px-3 py-3',
+        selected && 'bg-blue-100'
+      )}
+    >
+      {tabId === 'all' ? (
+        <span className={cn('font-semibold', selected && 'text-blue')}>
+          All
+        </span>
+      ) : (
+        <em-emoji shortcodes={tabId} size="1.3em" />
+      )}
+      <span className={cn('ml-2', selected && 'font-semibold text-blue')}>
+        {count}
+      </span>
+    </div>
+  );
+}
+
+interface ReactionsDisplayProps {
+  feels: Record<string, string>;
+  className?: string;
+}
+
+export default function ReactionsWidget({
+  feels,
+  className,
+}: ReactionsDisplayProps) {
+  const [selectedTab, setSelectedTab] = useState('all');
+
+  const feelCount = useMemo(() => Object.keys(feels).length, [feels]);
+  const feelsByShortcode = useMemo(() => _.invertBy(feels), [feels]);
+  const allFeels = useMemo(
+    () => Object.entries(feels).sort(byShortcode),
+    [feels]
+  );
+  const allShortcodes = useMemo(
+    () => Object.entries(feelsByShortcode).map(([code]) => code),
+    [feelsByShortcode]
+  );
```

Comment:
The memoization here could be unnecessary overhead (see https://blog.isquaredsoftware.com/2020/05/blogged-answers-a-mostly-complete-guide-to-react-rendering-behavior/#memoize-everything)

Reply:
All of these involve iterating over or sorting the list of reactions which is why they were memoized.

It is a bit of a toss up for now since that list is usually quite small. I think we're okay leaving this in though since the component is rendered infrequently and any overhead incurred won't affect app performance more broadly.